src/realm/util/websocket.cpp

```diff
@@ -678,14 +678,20 @@ class WebSocket {
 
         auto handler = [this](std::error_code ec, size_t) {
             // If the operation is aborted, the socket object may have been destroyed.
-            if (ec != util::error::operation_aborted) {
-                if (ec) {
-                    stop();
-                    m_config.websocket_write_error_handler(ec);
-                    return;
-                }
-                handle_write_message(); // Throws
+            // If the socket has been closed then we should continue to read from it until we've drained
+            // the receive buffer. Eventually we will either receive an in-band error message from the
+            // server about why we got disconnected or we'll receive ECONNRESET on the receive side as well.
+            if (ec == util::error::operation_aborted ||
+                ec == util::error::make_error_code(util::error::connection_reset) ||
+                ec == util::make_error_code(util::MiscExtErrors::end_of_input)) {
+                return;
             }
+            else if (ec) {
+                stop();
+                return m_config.websocket_write_error_handler(ec);
+            }
+
+            handle_write_message();
```

Comment:
Couple of nit picks, feel free to ignore:
- Superfluous `else`
- Problematic indentation, hard to see where the if-block starts. Personally, I would deal with this as follows:
```
bool closed = (ec == util::error::operation_aborted ||
               ec == util::error::make_error_code(util::error::connection_reset) ||
               ec == util::make_error_code(util::MiscExtErrors::end_of_input))
if (closed)
    return
```